src/main/java/com/flowingcode/vaadin/jsonmigration/JsonSerializer.java

```diff
@@ -0,0 +1,422 @@
+        JsonArray array = Json.createArray();
+        for (int i = 0; i < length; i++) {
+            array.set(i, toJson(Array.get(javaArray, i)));
+        }
+        return array;
+    }
+
+    private static Optional<JsonValue> tryToConvertToSimpleType(Object bean) {
+        if (bean instanceof String) {
+            return Optional.of(Json.create((String) bean));
+        }
+        if (bean instanceof Number) {
+            return Optional.of(Json.create(((Number) bean).doubleValue()));
+        }
+        if (bean instanceof Boolean) {
+            return Optional.of(Json.create((Boolean) bean));
+        }
+        if (bean instanceof Character) {
+            return Optional.of(Json.create(Character.toString((char) bean)));
+        }
+        if (bean instanceof Enum) {
+            return Optional.of(Json.create(((Enum<?>) bean).name()));
+        }
+        if (bean instanceof JsonValue) {
+            return Optional.of((JsonValue) bean);
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Converts a JsonValue to the corresponding Java object. The Java object
+     * can be a Java bean, {@link JsonSerializable} instance, String, wrapper of
+     * primitive types or an enum.
+     *
+     * @param type
+     *            the type of the Java object convert the json to
+     * @param json
+     *            the json representation of the object
+     * @param <T>
+     *            the resulting object type
+     *
+     * @return the deserialized object, or <code>null</code> if the input json
+     *         is <code>null</code>
+     */
+    public static <T> T toObject(Class<T> type, JsonValue json) {
+        return toObject(type, null, json);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T toObject(Class<T> type, Type genericType,
+            JsonValue json) {
+        if (json == null || json instanceof JsonNull) {
+            return null;
+        }
+
+        Optional<?> simpleType = tryToConvertFromSimpleType(type, json);
+        if (simpleType.isPresent()) {
+            return (T) simpleType.get();
+        }
+
+        if (Collection.class.isAssignableFrom(type)) {
+            return toCollection(type, genericType, json);
+        }
+
+        if (type.isRecord()) {
+            return toRecord(type, json);
+        }
+
+        T instance;
+        try {
+            instance = type.newInstance();
+        } catch (Exception e) {
+            throw new IllegalArgumentException(
+                    "Could not create an instance of type " + type
+                            + ". Make sure it contains a default public constructor and the class is accessible.",
+                    e);
+        }
+
+        try {
+//            if (instance instanceof JsonSerializable
+//                    && json instanceof JsonObject) {
+//                return type.cast(JsonSerializable.class.cast(instance)
+//                        .readJson((JsonObject) json));
+//            }
```

Comment:
Remove commented lines if no needed.

Reply:
These lines were active in the original Vaadin 24 implementation but have been commented out because JsonSerializable was removed in Vaadin 25. They remain here to indicate where this logic previously resided